### ISSUE TITLE
Ref resolver doesn't follow JSON Reference

### ISSUE THREAD

**Issue by yuri1969** [swagger-api/swagger-parser#1074]
Using **io.swagger.codegen.v3:swagger-codegen:3.0.7** it fails to resolve `$ref` aka _JSON Reference_. Calling `new OpenAPIV3Parser().read("src/openapi/openapi.yaml")` fails with:

```
java.lang.RuntimeException: Unable to load RELATIVE ref: ./schemas/index.yaml path: <path>/src/openapi
```

The API is not a Hello World so it is split to >30 files. Directory structure snippet:

```
openapi
├── components
│   ├── index.yaml
│   ├── parameters
│   │   └── index.yaml
│   ├── responses
│   │   └── index.yaml
│   └── schemas
│       ├── connection.yaml
│       ├── connectiondirection.yaml
│       ├── connectors
│       │   ├── connector.yaml
│       │    ...
│       │   └── impl
│       │       ├── foo.yaml
│       │        ...
│       │       └── bar.yaml
│       ├── index.yaml
│       ├── info.yaml
│        ...
│       ├── reports.yaml
│       ├── transaction.yaml
│       └── workflow.yaml
└── openapi.yaml
```

The issue is caused by a "nested" relative referencing:

```
/openapi.yaml
    $ref: "./components/index.yaml#foo"
/components/index.yaml
    foo:
       $ref: "./schemas/index.yaml#bar"
/components/schemas/index.yaml
    bar:
       ...
```

According to [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03): 

> Resolution is performed relative to the referring document.

This means the `./schemas/index.yaml` reference inside `/components/index.yaml` file shouldn't be resolved against the root directory but `/components` directory! On a first sight the `io.swagger.v3.parser.ResolverCache.loadRef()` operates with a static `parentDirectory` only.

This is a well-documented feature and [different parsers](https://github.com/swagger-api/swagger-ui/) work just fine.

**Comment by lorthirk** [comment]
Just wanted to +1 this issue. I originally thought it was related to `openapi-generator` but I think it belongs here. Please have a look at https://github.com/OpenAPITools/openapi-generator/issues/3848

**Comment by yuri1969** [comment]
FYI: I workaround this issue by flattening the whole directory structure a to single humongous YAML file using [Node-based swagger-parser](https://www.npmjs.com/package/swagger-parser) as a pre-processor in our CI pipeline.

**Comment by lorthirk** [comment]
Great hint. I'm in a Java based pipeline, but I'll see if I can manage to put up something similar. Thanks!

**Comment by lorthirk** [comment]
Unfortunately this is still an issue for me. I'm using `swagger-cli bundle` to generate one big `openapi.yaml`, but then using `swagger-codegen` to generate a Java Client outputs the code with a few errors and missing classes.

**Comment by yuri1969** [comment]
@lorthirk Unfortunately, I have no experience with `swagger-codegen`. Btw can the generated `openapi.yaml` file be linted correctly?

**Comment by lorthirk** [comment]
Yes, the resulting `openapi.yaml` is linted correctly. I think that `swagger-codegen` has a bug with code generation when there's a local reference (that contains `~1` instead of `/`).

**Comment by felixjung** [comment]
> Yes, the resulting `openapi.yaml` is linted correctly. I think that `swagger-codegen` has a bug with code generation when there's a local reference (that contains `~1` instead of `/`).

I can confirm that using a reference with escaped `/` (i.e., `~1`) causes relative references to files to break.

For example, I had the following reference in `index.yaml`:
```yaml
paths:
  /shipments:
    $ref: "./api/shipments.yaml#/~1shipments"
```

with `./api/shipments.yaml` starting like this:
```yaml
/shipments:
  get:
    summary: List all shipments
    operationId: listAllShipments
```

Removing the leading `/` in `shipments.yaml` and updating the reference resolved the issue.

**index.yaml**
```yaml
paths:
  /shipments:
    $ref: "./api/shipments.yaml#/shipments"
```

**shipments.yaml**
```yaml
shipments:
  get:
    summary: List all shipments
    operationId: listAllShipments
```

**Comment by apurvaSAP** [comment]
I am trying to reference a local file Money.json in another schema provisions.json using $ref as mentioned below:

**provisions.json**

`"responses": {
					"200": {
						"schema": {
							"$ref": "resources/json/Money.json#/definitions/Money"
						},`

**resources/json/Money.json**

`"definitions": {
    "Money": {
      "type": "object",
      "properties": {
        "amount": {
          "type": "number",
          "format": "bigdecimal",
          "description": "Amount"
        },
        "currency": {
          "type": "string",
          "description": "A unique internal identifier for the currency"
        }
      }
    }`

I want to customize the resource loading behaviour as per my requirement. For example I want to restrict the loading of remote URLs and allow local refs or interpret the ref URLs as per my application requirements. Is it possible to maybe register my custom reference resolver and loader. From the code I see RefUtils.java tries to load it but how can I customize it ?
